src/components/inputs/SearchSelect.jsx

```diff
@@ -0,0 +1,427 @@
+import React, { PropTypes, addons } from 'react/addons';
+const { update } = addons;
+import _ from 'lodash';
+import classNames from 'classnames';
+
+import InputMixin from '../../mixins/InputMixin';
+
+
+const KEY_CODES = {
+  TAB: 9,
+  ENTER: 13,
+  ARROW_UP: 38,
+  ARROW_DOWN: 40,
+};
+const DEFAULT_PLACEHOLDER = 'Search option';
+
+const optionPropType = PropTypes.shape({
+  id: PropTypes.string.isRequired,
+  label: PropTypes.string.isRequired,
+});
+
+const optionGroupOf = (_optionPropType) => PropTypes.oneOfType([
+  PropTypes.shape({
+    isGroup: PropTypes.bool.isRequired,
+    id: PropTypes.string.isRequired,
+    label: PropTypes.string.isRequired,
+    options: PropTypes.arrayOf(_optionPropType).isRequired,
+  }),
+  optionPropType,
+]);
+
+const OptionDefault = React.createClass({
+
+  displayName: 'Option',
+
+  propTypes: {
+    option: PropTypes.shape({
+      ...optionPropType,
+      label: PropTypes.string.isRequired,
+    }).isRequired,
+  },
+
+  render() {
+    let {
+      option: {label},
+      ...otherProps,
+    } = this.props;
+
+    return (
+      <div className="Option" {...otherProps}>
+        <span>{label}</span>
+      </div>
+    );
+  },
+
+});
+
+const Select = React.createClass({
+
+  displayName: 'Select',
+
+  mixins: [
+    InputMixin(optionPropType),
+  ],
+
+  propTypes: {
+    className: PropTypes.string,
+    placeHolder: PropTypes.string,
+    options: PropTypes.arrayOf(optionGroupOf(optionPropType)).isRequired,
+    optionRender: PropTypes.func,
+    filterOption: PropTypes.func, //By default filter option by their label.
+    hideGroupsWithNoMatch: PropTypes.bool,
+    valueLink: PropTypes.shape({
```

Comment:
The point is that InputMixin abstracts away the implementation details (ReactLink). This makes refactoring much easier. Also, the InputMixin API is far from complex and is very well documented.